src/tools/search/searXNGSearch.ts

```diff
@@ -0,0 +1,124 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ToolEmitter, Tool, ToolInput } from "@/tools/base.js";
+import { z } from "zod";
+import { Emitter } from "@/emitter/emitter.js";
+import { createURLParams } from "@/internals/fetcher.js";
+import { RunContext } from "@/context.js";
+import {
+  SearchToolOptions,
+  SearchToolOutput,
+  SearchToolResult,
+  SearchToolRunOptions,
+} from "@/tools/search/base.js";
+import { ValidationError } from "ajv";
+import { parseEnv } from "@/internals/env.js";
+
+export interface SearXNGToolOptions extends SearchToolOptions {
+  baseUrl?: string;
+  maxResults: number;
+}
+
+type SearXNGToolRunOptions = SearchToolRunOptions;
+
+export interface SearXNGToolResult extends SearchToolResult {}
+
+export class SearXNGToolOutput extends SearchToolOutput<SearXNGToolResult> {
+  static {
+    this.register();
+  }
+
+  createSnapshot() {
+    return {
+      results: this.results,
+    };
+  }
+
+  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>) {
+    Object.assign(this, snapshot);
+  }
+}
+
+export class SearXNGTool extends Tool<
+  SearXNGToolOutput,
+  SearXNGToolOptions,
+  SearXNGToolRunOptions
+> {
+  name = "Web Search";
+  description = `Search for online trends, news, current events, real-time information, or research topics.`;
+
+  public readonly emitter: ToolEmitter<ToolInput<this>, SearXNGToolOutput> = Emitter.root.child({
+    namespace: ["tool", "search", "searXNG"],
+    creator: this,
+  });
+
+  inputSchema() {
+    return z.object({
+      query: z.string().min(1).describe(`Web search query.`),
+    });
+  }
+
+  public constructor(options: SearXNGToolOptions = { maxResults: 10 }) {
+    super(options);
+
+    if (options.maxResults < 1 || options.maxResults > 100) {
+      throw new ValidationError([
+        {
+          message: "Property 'maxResults' must be between 1 and 100",
+          propertyName: "options.maxResults",
+        },
+      ]);
+    }
+  }
+
+  protected async _run(
+    input: ToolInput<this>,
+    _options: Partial<SearchToolRunOptions>,
+    run: RunContext<this>,
+  ) {
+    const params: URLSearchParams = ((input: ToolInput<this>) => {
+      return createURLParams({
+        q: input.query,
+        format: "json",
+      });
+    })(input);
+
+    const baseUrl = this.options.baseUrl || parseEnv("SEARXNG_BASE_URL", z.string());
+    const url = `${baseUrl}?${decodeURIComponent(params.toString())}`;
+    const response = await fetch(url, {
+      signal: run.signal,
+    });
+
+    if (!response.ok) {
+      throw new Error(
+        `The service request was unsuccessful. Code ${response.status} ${response.statusText}." `,
```

Comment:
I would add the whole body - the agent will not fix itself by just knowing the status code and status text. Pass `response.text()` data.